File: sdk/npm/src/providers/mohmal.ts
/**
 * Mohmal 临时邮箱渠道
 * 网站: mohmal.com
 * 基于 HTML 解析，使用 connect.sid session cookie 维持会话
 */
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "mohmal";
const BASE = "https://www.mohmal.com";

/** 从 HTML 中提取 data-email 属性的邮箱地址 */
const DATA_EMAIL_RE = /data-email="([^"]+)"/i;
/** 匹配收件箱表格中的邮件行 */
const TR_RE = /<tr[^>]*>[\s\S]*?<\/tr>/gi;
/** 提取邮件详情链接中的 ID */
const MSG_HREF_RE = /href="\/en\/message\/([^"]+)"/i;
/** 提取发件人单元格内容 */
const TD_FROM_RE = /<td[^>]*class="[^"]*sender[^"]*"[^>]*>([\s\S]*?)<\/td>/i;
/** 提取主题单元格内容 */
const TD_SUBJECT_RE =
  /<td[^>]*class="[^"]*subject[^"]*"[^>]*>([\s\S]*?)<\/td>/i;
/** 提取时间单元格内容 */
const TD_TIME_RE = /<td[^>]*class="[^"]*time[^"]*"[^>]*>([\s\S]*?)<\/td>/i;
/** 从邮件详情页提取邮件正文内容 */
const MAIL_BODY_RE =
  /<div[^>]*class="[^"]*mail-body[^"]*"[^>]*>([\s\S]*?)<\/div>/i;
/** 去除 HTML 标签 */
const TAG_RE = /<[^>]+>/g;

/** 浏览器模拟请求头 */
const PAGE_HEADERS: Record<string, string> = {
  Accept:
    "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
  "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
  "Cache-Control": "no-cache",
  DNT: "1",
  Pragma: "no-cache",
  "Upgrade-Insecure-Requests": "1",
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
};

/**
 * 从 set-cookie 响应头中提取所有 cookie 行
 */
function setCookieLines(headers: Headers): string[] {
  const h = headers as Headers & { getSetCookie?: () => string[] };
  if (typeof h.getSetCookie === "function") {
    return h.getSetCookie();
  }
  const one = headers.get("set-cookie");
  return one ? [one] : [];
}

/**
 * 解析 cookie 字符串为 Map
 */
function parseCookieMap(hdr: string): Map<string, string> {
  const m = new Map<string, string>();
  for (const part of hdr.split(";")) {
    const p = part.trim();
    if (!p || !p.includes("=")) continue;
    const i = p.indexOf("=");
    const k = p.slice(0, i).trim();
    const v = p.slice(i + 1).trim();
    if (k) m.set(k, v);
  }
  return m;
}

/**
 * 合并响应头中的 set-cookie 到已有 cookie 字符串
 */
function mergeCookieHeader(prev: string, headers: Headers): string {
  const m = parseCookieMap(prev);
  for (const line of setCookieLines(headers)) {
    const first = line.split(";")[0] ?? "";
    const i = first.indexOf("=");
    if (i <= 0) continue;
    const k = first.slice(0, i).trim();
    const v = first.slice(i + 1).trim();
    if (k) m.set(k, v);
  }
  return [...m.entries()]
    .sort((a, b) => a[0].localeCompare(b[0]))
    .map(([k, v]) => `${k}=${v}`)
    .join("; ");
}

/**
 * 去除 HTML 标签，返回纯文本
 */
function stripTags(s: string): string {
  return s.replace(TAG_RE, " ").replace(/\s+/g, " ").trim();
}

/**
 * HTML 实体反转义
 */
function htmlUnescape(s: string): string {
  return s
    .replace(/&amp;/g, "&")
    .replace(/&quot;/g, '"')
    .replace(/&#34;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&nbsp;/g, " ");
}

/** token 前缀，用于区分不同渠道的 token 格式 */
const TOK_PREFIX = "mohmal:";

/**
 * 编码 session 信息为 token 字符串
 * @param cookie - connect.sid 等 cookie 字符串
 */
function encodeToken(cookie: string): string {
  return TOK_PREFIX + Buffer.from(cookie, "utf8").toString("base64");
}

/**
 * 解码 token 字符串为 cookie
 */
function decodeToken(tok: string): string {
  if (!tok.startsWith(TOK_PREFIX)) {
    throw new Error("mohmal: invalid session token");
  }
  try {
    return Buffer.from(tok.slice(TOK_PREFIX.length), "base64").toString("utf8");
  } catch {
    throw new Error("mohmal: invalid session token");
  }
}

/**
 * 创建 Mohmal 临时邮箱
 *
 * 流程:
 * 1. GET /en/create/random 触发创建，服务端设置 connect.sid cookie
 * 2. 跟随重定向到 /en/inbox，从 HTML 中提取 data-email 属性获取邮箱地址
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  const createUrl = `${BASE}/en/create/random`;

  /* 第一步: 请求创建随机邮箱，手动处理重定向以收集 cookie */
  const r1 = await fetchWithTimeout(createUrl, {
    headers: { ...PAGE_HEADERS, Referer: `${BASE}/en` },
    redirect: "manual",
  });
  let cookieHdr = mergeCookieHeader("", r1.headers);

  /* 如果返回重定向，跟随到 /en/inbox */
  const location = r1.headers.get("location");
  if (location) {
    const inboxUrl = location.startsWith("http")
      ? location
      : `${BASE}${location}`;
    const r2 = await fetchWithTimeout(inboxUrl, {
      headers: { ...PAGE_HEADERS, Cookie: cookieHdr, Referer: createUrl },
    });
    if (!r2.ok) throw new Error(`mohmal: inbox page failed HTTP ${r2.status}`);
    cookieHdr = mergeCookieHeader(cookieHdr, r2.headers);
    const html = await r2.text();
    const email = extractEmail(html);
    const token = encodeToken(cookieHdr);
    return { channel: CHANNEL, email, token };
  }

  /* 非重定向情况，直接从响应体解析 */
  if (!r1.ok) throw new Error(`mohmal: create failed HTTP ${r1.status}`);
  const html = await r1.text();

  /* 检查是否已经是 inbox 页面 */
  const emailMatch = DATA_EMAIL_RE.exec(html);
  if (emailMatch?.[1]) {
    const email = htmlUnescape(emailMatch[1].trim());
    if (!email) throw new Error("mohmal: empty data-email");
    const token = encodeToken(cookieHdr);
    return { channel: CHANNEL, email, token };
  }

  /* 尝试请求 inbox 页面 */
  const r3 = await fetchWithTimeout(`${BASE}/en/inbox`, {
    headers: { ...PAGE_HEADERS, Cookie: cookieHdr, Referer: createUrl },
  });
  if (!r3.ok) throw new Error(`mohmal: inbox page failed HTTP ${r3.status}`);
  cookieHdr = mergeCookieHeader(cookieHdr, r3.headers);
  const inboxHtml = await r3.text();
  const email = extractEmail(inboxHtml);
  const token = encodeToken(cookieHdr);
  return { channel: CHANNEL, email, token };
}

/**
 * 从 inbox 页面 HTML 中提取邮箱地址
 */
function extractEmail(html: string): string {
  const m = DATA_EMAIL_RE.exec(html);
  if (!m?.[1]) throw new Error("mohmal: data-email not found in inbox page");
  const email = htmlUnescape(m[1].trim());
  if (!email) throw new Error("mohmal: empty data-email");
  return email;
}

/**
 * 从收件箱 HTML 表格行中解析邮件摘要信息
 */
function parseInboxRows(
  html: string,
): Array<{ id: string; from: string; subject: string; time: string }> {
  const rows: Array<{
    id: string;
    from: string;
    subject: string;
    time: string;
  }> = [];
  TR_RE.lastIndex = 0;
  let trMatch: RegExpExecArray | null;
  while ((trMatch = TR_RE.exec(html)) !== null) {
    const tr = trMatch[0];
    const hrefMatch = MSG_HREF_RE.exec(tr);
    if (!hrefMatch?.[1]) continue;
    const id = hrefMatch[1].trim();

    let from = "";
    const fromMatch = TD_FROM_RE.exec(tr);
    if (fromMatch?.[1]) from = htmlUnescape(stripTags(fromMatch[1]));

    let subject = "";
    const subjectMatch = TD_SUBJECT_RE.exec(tr);
    if (subjectMatch?.[1]) subject = htmlUnescape(stripTags(subjectMatch[1]));

    let time = "";
    const timeMatch = TD_TIME_RE.exec(tr);
    if (timeMatch?.[1]) time = htmlUnescape(stripTags(timeMatch[1]));

    rows.push({ id, from, subject, time });
  }
  return rows;
}

/**
 * 获取 Mohmal 邮箱的邮件列表
 *
 * 流程:
 * 1. GET /en/inbox 获取收件箱 HTML，解析邮件行
 * 2. 对每封邮件 GET /en/message/{id} 获取详情
 */
export async function getEmails(
  email: string,
  token: string,
): Promise<Email[]> {
  const cookieHdr = decodeToken(token);
  const inboxUrl = `${BASE}/en/inbox`;

  /* 获取收件箱页面 */
  const r = await fetchWithTimeout(inboxUrl, {
    headers: { ...PAGE_HEADERS, Cookie: cookieHdr, Referer: `${BASE}/en` },
  });
  if (!r.ok) throw new Error(`mohmal: inbox fetch failed HTTP ${r.status}`);
  const inboxHtml = await r.text();

  /* 解析邮件列表 */
  const rows = parseInboxRows(inboxHtml);
  const out: Email[] = [];
  for (const row of rows) {
    /* 获取邮件详情页 */
    const detailUrl = `${BASE}/en/message/${row.id}`;
    const rd = await fetchWithTimeout(detailUrl, {
      headers: { ...PAGE_HEADERS, Cookie: cookieHdr, Referer: inboxUrl },
    });
    if (!rd.ok) continue;
    const detailHtml = await rd.text();

    /* 提取邮件正文 */
    let htmlBody = "";
    const bodyMatch = MAIL_BODY_RE.exec(detailHtml);
    if (bodyMatch?.[1]) htmlBody = bodyMatch[1].trim();

    const raw: Record<string, unknown> = {
      id: row.id,
      from: row.from,
      to: email,
      subject: row.subject,
      date: row.time,
      html: htmlBody,
    };
    out.push(normalizeEmail(raw, email));
  }
  return out;
}

File: sdk/rust/src/providers/mohmal.rs
/*!
 * mohmal.com: connect.sid session cookie + HTML 收件箱
 * 通过 GET /en/create/random 创建邮箱，从 HTML 中提取 data-email 属性获取邮箱地址
 * 通过 GET /en/inbox 获取邮件列表，解析 #inbox-table 表格行
 * 通过 GET /en/message/{id} 获取邮件详情，从 HTML 中提取邮件内容
 */

use std::collections::BTreeMap;
use std::sync::LazyLock;

use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

use crate::config::{block_on, get_current_ua, http_client_no_cookie_jar};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};

const ORIGIN: &str = "https://www.mohmal.com";
/// token 前缀，用于序列化 / 反序列化会话信息
const TOK_PREFIX: &str = "mohmal1:";

/// 序列化到 token 中的会话信息
#[derive(Serialize, Deserialize)]
struct MohmalSess {
    /// Cookie 字符串
    c: String,
}

/// 匹配 data-email 属性
static DATA_EMAIL_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r#"data-email="([^"]+)""#).expect("re"));

/// 匹配收件箱表格行中的邮件链接
static MSG_LINK_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r#"href="/en/message/([^"]+)""#).expect("re"));

/// 匹配表格行
static TR_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r#"(?is)<tr[^>]*>([\s\S]*?)</tr>"#).expect("re"));

/// 匹配表格单元格
static TD_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r#"(?is)<td[^>]*>([\s\S]*?)</td>"#).expect("re"));

/// 匹配 HTML 标签
static TAG_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r#"<[^>]+>"#).expect("re"));

/// 匹配邮件详情页内容区域（开标签，用于栈式解析）
static MSG_BODY_OPEN_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#"(?is)<div[^>]*class="[^"]*mail-content[^"]*"[^>]*>"#).expect("re")
});

/// 匹配邮件详情页的主题
static MSG_SUBJECT_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#"(?is)<h4[^>]*class="[^"]*subject[^"]*"[^>]*>([\s\S]*?)</h4>"#).expect("re")
});

/// 匹配邮件详情页的发件人
static MSG_FROM_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#"(?is)<span[^>]*class="[^"]*from[^"]*"[^>]*>([\s\S]*?)</span>"#).expect("re")
});

/// 匹配邮件详情页的日期
static MSG_DATE_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#"(?is)<span[^>]*class="[^"]*date[^"]*"[^>]*>([\s\S]*?)</span>"#).expect("re")
});

/// 匹配 iframe src 中的 HTML 内容地址
static IFRAME_SRC_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r#"(?is)<iframe[^>]*src="([^"]+)"[^>]*>"#).expect("re"));

/// 使用栈式深度匹配提取 mail-content div 的完整内部 HTML，
/// 避免非贪婪正则在嵌套 div 时截断正文。
fn extract_mail_body(page: &str) -> String {
    let Some(m) = MSG_BODY_OPEN_RE.find(page) else {
        return String::new();
    };
    let start = m.end();
    let mut pos = start;
    let mut depth = 1usize;
    while pos < page.len() && depth > 0 {
        let rest = &page[pos..];
        let next_open = rest.find("<div").map(|i| pos + i);
        let next_close = rest.find("</div>").map(|i| pos + i);
        match (next_open, next_close) {
            (_, None) => break,
            (Some(no), Some(nc)) if no < nc => {
                depth += 1;
                pos = no + 4;
            }
            (_, Some(nc)) => {
                depth -= 1;
                if depth == 0 {
                    return page[start..nc].trim().to_string();
                }
                pos = nc + 6;
            }
        }
    }
    String::new()
}

/// 解析 Cookie 头为 key=value 映射
fn parse_cookie_header(hdr: &str) -> BTreeMap<String, String> {
    let mut m = BTreeMap::new();
    for part in hdr.split(';') {
        let part = part.trim();
        if part.is_empty() {
            continue;
        }
        if let Some(i) = part.find('=') {
            let k = part[..i].trim();
            let v = part[i + 1..].trim();
            if !k.is_empty() {
                m.insert(k.to_string(), v.to_string());
            }
        }
    }
    m
}

/// 将 Cookie 映射转换回 Cookie 头字符串
fn cookie_header_from_map(m: &BTreeMap<String, String>) -> String {
    m.iter()
        .map(|(k, v)| format!("{}={}", k, v))
        .collect::<Vec<_>>()
        .join("; ")
}

/// 合并 Set-Cookie 响应头到已有 Cookie 中
fn merge_set_cookies(hdr: &str, headers: &wreq::header::HeaderMap) -> String {
    let mut m = parse_cookie_header(hdr);
    for val in headers.get_all("set-cookie") {
        let Ok(s) = val.to_str() else {
            continue;
        };
        let first = s.split(';').next().unwrap_or("");
        if let Some(i) = first.find('=') {
            let k = first[..i].trim();
            let v = first[i + 1..].trim();
            if !k.is_empty() {
                m.insert(k.to_string(), v.to_string());
            }
        }
    }
    cookie_header_from_map(&m)
}

/// 编码会话信息为 token 字符串
fn encode_sess(s: &MohmalSess) -> Result<String, String> {
    let b = serde_json::to_vec(s).map_err(|e| e.to_string())?;
    Ok(format!(
        "{}{}",
        TOK_PREFIX,
        base64::Engine::encode(&base64::engine::general_purpose::STANDARD, b)
    ))
}

/// 从 token 字符串解码会话信息
fn decode_sess(tok: &str) -> Result<MohmalSess, String> {
    if !tok.starts_with(TOK_PREFIX) {
        return Err("mohmal: token 格式无效".into());
    }
    let raw = base64::Engine::decode(
        &base64::engine::general_purpose::STANDARD,
        tok.trim_start_matches(TOK_PREFIX),
    )
    .map_err(|_| "mohmal: token 解码失败".to_string())?;
    let s: MohmalSess =
        serde_json::from_slice(&raw).map_err(|_| "mohmal: token 反序列化失败".to_string())?;
    if s.c.is_empty() {
        return Err("mohmal: token 中 cookie 为空".into());
    }
    Ok(s)
}

/// 去除 HTML 标签
fn strip_tags(s: &str) -> String {
    TAG_RE.replace_all(s, " ").to_string()
}

/// 简单 HTML 实体反转义
fn light_unescape(s: &str) -> String {
    let t = s.replace("&amp;", "\u{fffd}");
    t.replace("&quot;", "\"")
        .replace("&#34;", "\"")
        .replace("&#39;", "'")
        .replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&nbsp;", " ")
        .replace('\u{fffd}', "&")
}

/// 构建浏览器请求头
fn page_headers(referer: &str) -> Vec<(&'static str, String)> {
    vec![
        ("User-Agent", get_current_ua().to_string()),
        (
            "Accept",
            "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8"
                .to_string(),
        ),
        (
            "Accept-Language",
            "en-US,en;q=0.9".to_string(),
        ),
        ("Cache-Control", "no-cache".to_string()),
        ("DNT", "1".to_string()),
        ("Pragma", "no-cache".to_string()),
        ("Referer", referer.to_string()),
        ("Upgrade-Insecure-Requests", "1".to_string()),
    ]
}

/// 创建临时邮箱
/// 1. GET /en/create/random，跟随重定向到 /en/inbox，获取 connect.sid
/// 2. 从 /en/inbox 页面 HTML 中提取 data-email 属性获取邮箱地址
pub fn generate_email() -> Result<EmailInfo, String> {
    let create_url = format!("{}/en/create/random", ORIGIN);
    let inbox_url = format!("{}/en/inbox", ORIGIN);

    block_on(async {
        let client = http_client_no_cookie_jar();

        // 构建不跟随重定向的客户端，手动跟踪 cookie
        let no_redirect_client = wreq::Client::builder()
            .redirect(wreq::redirect::Policy::none())
            .build()
            .map_err(|e| format!("mohmal: 创建客户端失败: {}", e))?;

        // 第一步：GET /en/create/random（不跟随重定向，获取 connect.sid cookie）
        let mut req = no_redirect_client.get(&create_url);
        for (k, v) in page_headers(ORIGIN) {
            req = req.header(k, v);
        }
        let resp = req
            .send()
            .await
            .map_err(|e| format!("mohmal create: {}", e))?;
        let mut cookie_hdr = merge_set_cookies("", resp.headers());

        // 检查是否存在 connect.sid cookie
        if !parse_cookie_header(&cookie_hdr).contains_key("connect.sid") {
            // 可能服务端直接返回了 200 而非重定向，尝试从响应体提取
            let status = resp.status();
            if status.is_redirection() {
                // 手动跟随重定向到 /en/inbox
                let location = resp
                    .headers()
                    .get("location")
                    .and_then(|v| v.to_str().ok())
                    .unwrap_or("/en/inbox");
                let redirect_url = if location.starts_with("http") {
                    location.to_string()
                } else {
                    format!("{}{}", ORIGIN, location)
                };
                let mut req2 = no_redirect_client.get(&redirect_url);
                for (k, v) in page_headers(&create_url) {
                    req2 = req2.header(k, v);
                }
                req2 = req2.header("Cookie", &cookie_hdr);
                let resp2 = req2
                    .send()
                    .await
                    .map_err(|e| format!("mohmal redirect: {}", e))?;
                cookie_hdr = merge_set_cookies(&cookie_hdr, resp2.headers());
            }
        }

        if !parse_cookie_header(&cookie_hdr).contains_key("connect.sid") {
            return Err("mohmal: 未获取到 connect.sid cookie".into());
        }

        // 第二步：GET /en/inbox 获取邮箱地址
        let mut req3 = client.get(&inbox_url);
        for (k, v) in page_headers(&create_url) {
            req3 = req3.header(k, v);
        }
        req3 = req3.header("Cookie", &cookie_hdr);
        let resp3 = req3
            .send()
            .await
            .map_err(|e| format!("mohmal inbox: {}", e))?;
        if !resp3.status().is_success() {
            return Err(format!("mohmal inbox: {}", resp3.status()));
        }
        cookie_hdr = merge_set_cookies(&cookie_hdr, resp3.headers());
        let html = resp3.text().await.map_err(|e| e.to_string())?;

        // 从 HTML 中提取 data-email 属性
        let email = DATA_EMAIL_RE
            .captures(&html)
            .and_then(|c| c.get(1))
            .map(|m| light_unescape(m.as_str().trim()))
            .ok_or_else(|| "mohmal: 未找到 data-email 属性".to_string())?;

        if email.is_empty() || !email.contains('@') {
            return Err("mohmal: 提取的邮箱地址无效".into());
        }

        let tok = encode_sess(&MohmalSess { c: cookie_hdr })?;

        Ok(EmailInfo {
            channel: Channel::Mohmal,
            email,
            token: Some(tok),
            expires_at: None,
            created_at: None,
        })
    })
}

/// 获取邮件列表
/// 1. GET /en/inbox 获取收件箱 HTML
/// 2. 解析 #inbox-table 中的邮件链接和基本信息
/// 3. 对每封邮件 GET /en/message/{id} 获取详细内容
pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let sess = decode_sess(token)?;
    let inbox_url = format!("{}/en/inbox", ORIGIN);
    let cookie_hdr = &sess.c;

    block_on(async {
        let client = http_client_no_cookie_jar();

        // 获取收件箱页面
        let mut req = client.get(&inbox_url);
        for (k, v) in page_headers(ORIGIN) {
            req = req.header(k, v);
        }
        req = req.header("Cookie", cookie_hdr);
        let resp = req
            .send()
            .await
            .map_err(|e| format!("mohmal inbox: {}", e))?;
        if !resp.status().is_success() {
            return Err(format!("mohmal inbox: {}", resp.status()));
        }
        let html = resp.text().await.map_err(|e| e.to_string())?;

        // 解析邮件列表：提取每行的消息 ID 和基本信息
        let mail_entries = parse_inbox_table(&html);
        let mut out = Vec::with_capacity(mail_entries.len());

        for entry in &mail_entries {
            // 获取每封邮件的详情页
            let detail_url = format!("{}/en/message/{}", ORIGIN, entry.id);
            let mut reqd = client.get(&detail_url);
            for (k, v) in page_headers(&inbox_url) {
                reqd = reqd.header(k, v);
            }
            reqd = reqd.header("Cookie", cookie_hdr);
            let respd = match reqd.send().await {
                Ok(r) => r,
                Err(_) => continue,
            };
            if !respd.status().is_success() {
                continue;
            }
            let page = match respd.text().await {
                Ok(t) => t,
                Err(_) => continue,
            };

            let raw = parse_detail_page(&page, entry, email);
            out.push(normalize_email(&raw, email));
        }

        Ok(out)
    })
}

/// 收件箱中解析出的邮件条目
struct InboxEntry {
    /// 邮件 ID
    id: String,
    /// 发件人
    from: String,
    /// 主题
    subject: String,
    /// 时间
    date: String,
}

/// 从收件箱 HTML 解析邮件列表条目
fn parse_inbox_table(html: &str) -> Vec<InboxEntry> {
    let mut entries = Vec::new();
    let mut seen = std::collections::HashSet::new();

    for tr_cap in TR_RE.captures_iter(html) {
        let tr_content = tr_cap.get(1).unwrap().as_str();

        // 提取消息 ID
        let msg_id = match MSG_LINK_RE.captures(tr_content) {
            Some(cap) => cap.get(1).unwrap().as_str().to_string(),
            None => continue,
        };

        if !seen.insert(msg_id.clone()) {
            continue;
        }

        // 提取所有 td 单元格
        let tds: Vec<String> = TD_RE
            .captures_iter(tr_content)
            .map(|c| {
                let raw = c.get(1).unwrap().as_str();
                light_unescape(strip_tags(raw).trim())
            })
            .collect();

        // 表格行结构：发件人, 主题, 时间
        let from = tds.first().cloned().unwrap_or_default();
        let subject = tds.get(1).cloned().unwrap_or_default();
        let date = tds.get(2).cloned().unwrap_or_default();

        entries.push(InboxEntry {
            id: msg_id,
            from,
            subject,
            date,
        });
    }

    entries
}

/// 从邮件详情页 HTML 解析邮件内容
fn parse_detail_page(page: &str, entry: &InboxEntry, recipient: &str) -> Value {
    // 尝试从详情页提取更精确的字段，回退到列表中的基本信息
    let from = MSG_FROM_RE
        .captures(page)
        .and_then(|c| c.get(1))
        .map(|m| light_unescape(strip_tags(m.as_str()).trim()))
        .filter(|s| !s.is_empty())
        .unwrap_or_else(|| entry.from.clone());

    let subject = MSG_SUBJECT_RE
        .captures(page)
        .and_then(|c| c.get(1))
        .map(|m| light_unescape(strip_tags(m.as_str()).trim()))
        .filter(|s| !s.is_empty())
        .unwrap_or_else(|| entry.subject.clone());

    let date = MSG_DATE_RE
        .captures(page)
        .and_then(|c| c.get(1))
        .map(|m| light_unescape(strip_tags(m.as_str()).trim()))
        .filter(|s| !s.is_empty())
        .unwrap_or_else(|| entry.date.clone());

    // 提取邮件正文 HTML：使用栈式解析避免嵌套 div 截断
    let html_body = extract_mail_body(page);
    let html_body = if html_body.is_empty() {
        // 回退：检查是否有 iframe src
        if let Some(iframe_cap) = IFRAME_SRC_RE.captures(page) {
            let src = iframe_cap.get(1).unwrap().as_str();
            format!("<p>Content available at: {}</p>", src)
        } else {
            String::new()
        }
    } else {
        html_body
    };

    json!({
        "id": entry.id,
        "to": recipient,
        "from": from,
        "subject": subject,
        "date": date,
        "html": html_body,
    })
}
